### PR TITLE
Document processing v2

### PR DESCRIPTION
**Pull request by Siegrift**
Relates to https://github.com/api3dao/commons/issues/27

## Rationale

Read https://github.com/api3dao/commons/pull/32

### CODE REVIEW

File: docs/reference/ois/next/specification.md [hunk 1]
```diff
@@ -594,8 +589,16 @@ corresponding operation parameter.-->
 
 ### 5.9. `preProcessingSpecifications` \*
 
-(Optional) Defines the preprocessing code that can be used to modify the
-endpoint parameter before making the API request defined by an Airnode endpoint.
+::: deprecation warning
```
Comment (Siegrift) [inline docs/reference/ois/next/specification.md:592]: We are not looking to remove it just yet, but we want encourage users to migrate so I wanted to give some encouragement message.

File: docs/reference/ois/next/processing.md [hunk 2]
```diff
@@ -19,6 +19,22 @@ tags:
...
+
+- Authentication
+- Data transformation
+- Data aggregation
+- Data validation
+
+## Processing versions
+
+OIS specification has two versions of pre/post processing. Both versions serve
+the same use cases, but the second version is more flexible and convenient.
```
Comment (Siegrift) [inline docs/reference/ois/next/processing.md:33]: By more flexible I mean we can implement the use cases mentioned in https://github.com/api3dao/commons/issues/27

By convenient I mean that processing is just a function now.

File: docs/reference/ois/next/processing.md [hunk 3]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+  parameters. For example, if there was a parameter named `myParameter` defined
+  in the `endpoints[n].parameters` array, its value could be accessed using
+  `endpointParameters.myParameter` within pre-processing snippet.
+
+The output of the post-processing snippet is an object with the following
+fields:
+
+- `apiCallResponse` - The post-processed API call response. This is used to
+  encode the response values defined by reservedParameters.
+- `timestamp` - (Optional) The timestamp of the API call response. Use this if
```
Comment (Siegrift) [inline docs/reference/ois/next/processing.md:160]: fyi: this is the support for the new use case mentioned in https://github.com/api3dao/commons/issues/27

File: docs/reference/ois/next/specification.md [hunk 4]
```diff
@@ -375,25 +379,16 @@ node.
...
-      }
-    ],
-    "postProcessingSpecifications": [
-      {
-        "environment": "Node",
-        "value": "const output = Math.round(input.price * 1000);",
-        "timeoutMs": "5000"
-      }
-    ]
+    "preProcessingSpecificationV2": {
```
Comment (Siegrift) [inline docs/reference/ois/next/specification.md:382]: I changed the example here to make it more obvious how v2 is used, but I didn't migrate all of the snippets in the docs (because v1 still works).

File: docs/reference/ois/next/processing.md [hunk 5]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+#### Input and Output
+
+The processing snippet is a function which receives payload as an argument. The
+return value of the function is treated as a processing result. Apart from the
+payload argument, you can use most Node.js built-in modules.
+
+The payload argument for pre-processing is an object with the following
+properties:
+
+- `apiCallParameters` - The API call parameters with the exception of reserved
```
Comment (bbenligiray) [inline docs/reference/ois/next/processing.md:136]: Was this supposed to be `endpointParameters`?

File: docs/reference/ois/next/processing.md [hunk 6]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+  parameters. For example, if there was a parameter named `myParameter` defined
+  in the `endpoints[n].parameters` array, its value could be accessed using
+  `endpointParameters.myParameter` within pre-processing snippet.
+
+The output of the pre-processing snippet is an object with the following fields:
+
+- `apiCallParameters` - The pre-processed API call parameters. These are used to
```
Comment (bbenligiray) [inline docs/reference/ois/next/processing.md:143]: `operationParameters` should work here if we want to limit the vocabulary

File: docs/reference/ois/next/processing.md [hunk 7]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+  make the API call.
+
+The payload argument for post-processing is an object with the following
+properties:
+
+- `apiCallResponse` - The API call response.
+- `endpointParameters` - The API call parameters with the exception of reserved
```
Comment (bbenligiray) [inline docs/reference/ois/next/processing.md:150]: These are not necessarily the parameters that the API is called with so you shouldn't call them as such. So the description should be `"The endpoints parameters with the exception..."`

File: docs/reference/ois/next/processing.md [hunk 8]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+  parameters. For example, if there was a parameter named `myParameter` defined
+  in the `endpoints[n].parameters` array, its value could be accessed using
+  `endpointParameters.myParameter` within pre-processing snippet.
+
+The output of the post-processing snippet is an object with the following
+fields:
+
+- `apiCallResponse` - The post-processed API call response. This is used to
```
Comment (bbenligiray) [inline docs/reference/ois/next/processing.md:158]: This is not the API call response, it's the post-processed version of it, so this feels misnamed to me

File: docs/reference/ois/next/processing.md [hunk 9]
```diff
@@ -245,4 +333,24 @@ endpoints: [
 ]
 ```
 
+## Example #3
+
+One of the possible use cases for post-processing would be to use override the
```
Comment (bbenligiray) [inline docs/reference/ois/next/processing.md:338]: Will this also work for RRP?

File: docs/reference/ois/next/processing.md (hunk 6 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:143]: What I like though is that you receive `apiCallParameters` in the payload of the function and you can return the same thing. It's wierd when they have a different name.

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:158]: My reasoning is that you can return the same response that you receive in the function, similarly to doing `output = input` in the v1 processing.

File: docs/reference/ois/next/processing.md (hunk 7 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:150]: > These are not necessarily the parameters that the API is called with

I don't understand what you mean.

As a note, these should be called the same as the endpoint parameters in the pre-processing.

File: docs/reference/ois/next/processing.md (hunk 9 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:338]: No, the timestamp is only used for signed data responses [here](https://github.com/api3dao/airnode/blob/02d15254dfc431b4c0aa06b36f64953c550849fd/packages/airnode-node/src/api/index.ts#L277)

File: docs/reference/ois/next/processing.md (hunk 5 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:136]: Yes please. There is a significant semantic difference and `endpointParameters` is much more accurate.

File: docs/reference/ois/next/processing.md (hunk 6 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:143]: The semantic problem with mutating something is that you have two versions of the same thing that you call with the same name. For example, if the input is `endpointParameters`, calling the output `preprocessedEndpointParameters` would be more accurate than calling it `endpointParameters`. That being said, these endpoint parameters will be used as API operation parameters next, so they could be called that. `apiCallParameters` also works here, I'm just saying that since we're already talking about API operations a lot in the context of OIS (and in contrast "API call" is not introduced as a concept), it may be better to reuse that instead of adding a term.

File: docs/reference/ois/next/processing.md (hunk 7 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:150]: For example, there can be an endpoint parameter that only specifies which code branch is to be taken in post-processing. That would be an endpoint parameter but not an API call parameter.

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:158]: In v1, `output` is named differently than `input`. The equivalent of `output = input` would be `apiCallResponse = postProcessedApiCallResponse`.

File: docs/reference/ois/next/processing.md (hunk 9 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:338]: Any particular reason

**Comment by bbenligiray** [comment]
One thing to note is that it's possible to specify an endpoint with no API call, so `apiCallParameters`, `apiCallResponse`, etc. are not suitable. A more generic naming would be more compatible with other kinds of operations (other than API calls) or more modular stuff like chained API calls, etc. that may be implemented in the future.

File: docs/reference/ois/next/processing.md (hunk 5 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:136]: > One thing to note is that it's possible to specify an endpoint with no API call, so apiCallParameters, apiCallResponse, etc. are not suitable.

This is also a big downside for calling it apiCallParameters. I'll revert.

File: docs/reference/ois/next/processing.md (hunk 6 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:143]: I dislike the `apiCallParameters` now and I slightly prefer `operationParameters` over `preprocessedEndpointParameters `.

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:158]: It would actually be `postProcessedApiCallResponse  = apiCallResponse`, but since we have function what you want is `return apiCallResponse` but we can't do that, because we also want to (optionally) return timestamp so we need to do it like `return {apiCallResponse}`.

File: docs/reference/ois/next/processing.md (hunk 9 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:338]: To be honest I am not sure what you mean. The RRPv0 doesn't have use timestamp in its [fulfill function](https://github.com/api3dao/airnode/blob/processing-v2/packages/airnode-protocol/contracts/rrp/AirnodeRrpV0.sol#L226).

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:158]: I was thinking a bit more about:

> This is not the API call response, it's the post-processed version of it, so this feels misnamed to me

Since Airnode/Pusher does the processing and its deployed by the API provider themselves, to me the API response is actually a combination of API call + post-processing. However as you've said, there is possibility of skipping API call in which case the apiCallResponse (or apiResponse) don't make sense and I can't think of a good name for this.

One possibility I had in mind would be to call it `apiCallResponse` (or `apiResponse`) and when we skip a call we would pass `undefined`, which kinda makes sense. Currently, we [use the pre-processing output as API response](https://github.com/api3dao/airnode/blob/02d15254dfc431b4c0aa06b36f64953c550849fd/packages/airnode-node/src/api/index.ts#L323) so with this change the post-processing would be a must have when skipping an API call.

File: docs/reference/ois/next/specification.md (hunk 1 above)
Comment (dcroote) [reply docs/reference/ois/next/specification.md:592]: This isn't rendering because `::: deprecation` isn't a container type. I think you want:
```suggestion
::: warning Deprecation
```

File: docs/reference/ois/next/processing.md [hunk 10]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+understand pre/post processing.
+
+#### Input and Output
+
+The processing snippet is a function which receives payload as an argument. The
+return value of the function is treated as a processing result. Apart from the
+payload argument, you can use most Node.js built-in modules.
+
+The payload argument for pre-processing is an object with the following
+properties:
```
Comment (dcroote) [inline docs/reference/ois/next/processing.md:134]: "propert**ies**" plural, but there is only `endpointParameters`?

File: docs/reference/ois/next/processing.md [hunk 11]
```diff
@@ -77,14 +93,84 @@ post-processing.
...
+
+- `endpointParameters` - The endpoint parameters with the exception of reserved
+  parameters. For example, if there was a parameter named `myParameter` defined
+  in the `endpoints[n].parameters` array, its value could be accessed using
+  `endpointParameters.myParameter` within pre-processing snippet.
+
+The output of the pre-processing snippet is an object with the following fields:
```
Comment (dcroote) [inline docs/reference/ois/next/processing.md:141]: "field**s**" plural, but there is only `endpointParameters`?

File: docs/reference/ois/next/specification.md [hunk 12]
```diff
@@ -610,21 +613,28 @@ additional details.
...
     // Run for 5 seconds maximum
-    "timeoutMs": "5000"
+    "timeoutMs": 5000
   }
 ]
 ```
 
 ### 5.10. `postProcessingSpecifications` \*
 
+::: deprecation warning
```
Comment (dcroote) [inline docs/reference/ois/next/specification.md:631]: Same comment:
```suggestion
::: warning Deprecation
```

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (dcroote) [reply docs/reference/ois/next/processing.md:158]: The more technically correct output might be `postProcessedApiCallResponse`, but I don't disagree with this either:
> to me the API response is actually a combination of API call + post-processing

Post-processing as an "inplace" update to the API call response is effectively what we're doing i.e. don't use the actual API call response, use this instead.

Having a different return variable would also mean that modifying only the timestamp would eliminate property value shorthand:

```ts
"async ({ apiCallResponse }) => { return { postProcessedApiCallResponse: apiCallResponse, timestamp: apiCallResponse.ProviderTimestamp }; }"
```

tl;dr I'd be fine leaving it as `apiCallResponse`.

File: docs/reference/ois/next/specification.md (hunk 1 above)
Comment (Siegrift) [reply docs/reference/ois/next/specification.md:592]: Thanks for noticing.

File: docs/reference/ois/next/processing.md (hunk 10 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:134]: I didn't mind for two reasons:
- It's future proof
- It's not a "cohesive" text. The sentence says that the object has certain properties and follows up with a list of them. The list has just a single item.

That said, youre suggestion is more accurate so I'll use that.

File: docs/reference/ois/next/processing.md (hunk 11 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:141]: Will use property (so that we are consistent and do not use both field and property).

File: docs/reference/ois/next/processing.md (hunk 6 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md]: All options are kinda bad so I'm fine with both the input and the output being called `endpointParameters` here

Comment (bbenligiray) [reply docs/reference/ois/next/processing.md]: After reading on a bit, it didn't feel clear if `endpointParameters` in post-processing are raw endpoint parameters or pre-processed endpoint parameters, which is another downside of calling the output of this `endpointParameters`. As a counterargument to that, if we call the output of this `preprocessedEndpointParameters`, it will appear like endpoint parameters in OIS (https://docs.api3.org/reference/ois/latest/specification.html#_5-5-parameters) are mapped to raw endpoint parameters and not pre-processed endpoint parameters.
I guess simply mentioning in the post-processing docs that `endpointParameters` are raw/un-preprocessed is the best solution to this.

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:159]: I personally see pre/post-processing as a part of the endpoint specs and not API functionality. However, leaving this as `apiCallResponse` would be consistent with having the output of pre-processing be named `endpointParameters` (which I had agreed on above).
One final nit, one can use post-processing without an API call (https://docs.api3.org/reference/ois/latest/processing.html#skip-the-api-call), in which case this naming would be off. The previous `input`/`output` names worked nicely with this kind of ambiguity. An easy fix would be to call this `response` instead of `apiCallResponse`.

File: docs/reference/ois/next/processing.md (hunk 9 above)
Comment (bbenligiray) [reply docs/reference/ois/next/processing.md:339]: At this point I forgot how RRP works :skull:

**Comment by bbenligiray** [comment]
I left a few small but imo somewhat important comments. If you respond and need me to respond, feel free to poke me with a DM.

File: docs/reference/ois/next/processing.md (hunk 6 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md]: Leaving them as `endpointParameters` and clarifying the `endpointParameters` in the post-processing section.

File: docs/reference/ois/next/processing.md (hunk 8 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:159]: Calling it `response` and clarifying this in the docs.


As a last thing, @dcroote @bbenligiray What do you think about:
> One possibility I had in mind would be to call it apiCallResponse (or apiResponse) and when we skip a call we would pass undefined, which kinda makes sense. Currently, we [use the pre-processing output as API response](https://github.com/api3dao/airnode/blob/02d15254dfc431b4c0aa06b36f64953c550849fd/packages/airnode-node/src/api/index.ts#L323) so with this change the post-processing would be a must have when skipping an API call.

Comment (bbenligiray) [reply docs/reference/ois/next/processing.md]: Then pre-processing (v2) while the API call is skipped is entirely useless so it shouldn't be allowed imo.
Also, you should disallow using different versions (v1/v2) of pre and post processing unless you haven't already.

Comment (dcroote) [reply docs/reference/ois/next/processing.md]: For v1 we allow either pre or post-processing when an API call is skipped: https://docs.api3.org/reference/ois/latest/processing.html#skip-the-api-call
>Instead of calling an API, Airnode uses the output of preProcessingSpecifications, postProcessingSpecifications, or both.


But I agree with Burak that pre doesn't really make sense if the API call is skipped.

Comment (Siegrift) [reply docs/reference/ois/next/processing.md]: Ok. Resolving the discussion and going with the current state. I think everything should now be as we agreed.

**Comment by Siegrift** [comment]
I think the discussion is now done. @dcroote @wkande can you take a final look and approve?

File: docs/reference/ois/next/processing.md [hunk 13]
```diff
@@ -77,14 +94,89 @@ post-processing.
...
+The output of the post-processing snippet is an object with the following
+properties:
+
+- `response` - The post-processed API call response (or post-processed result of
+  pre-processing snippet in case of skipping an API call). This is used to
+  encode the response values defined by reserved parameters.
+- `timestamp` - (Optional) The timestamp of the API call response. Use this if
+  you want Airnode to use a specific timestamp (instead of a current time at
+  request processing) when using the [signed data
+  gateway](https://docs.api3.org/reference/airnode/latest/understand/http-gateways.html#http-signed-data-gateway).
```
Comment (dcroote) [inline docs/reference/ois/next/processing.md:169]: Local link:
```suggestion
  gateway](/reference/airnode/latest/understand/http-gateways.md#http-signed-data-gateway).
```

**Review by dcroote — APPROVED** [review]
Only minor / nits, otherwise LGTM 👍

File: docs/reference/ois/next/processing.md [hunk 14]
```diff
@@ -77,14 +94,89 @@ post-processing.
...
+and
+[`postProcessingSpecificationV2`](/reference/ois/next/specification.md#_5-12-postprocessingspecificationv2).
+Snippets for both specifications follow this schema:
+
+- `environment` - Currently only possible value is `Node`. This options
+  interprets the code as JavaScript and executes it in Node.js. The function can
+  be also asynchronous (async/await is supported as well). The processing
+  implementation will wait for the function to resolve.
+- `value` - The processing code written as a string.
+- `timeoutMs` - The maximum timeout that this snippet can run. In case the
```
[the remark this answers is not in the document]
Comment (Siegrift) [reply docs/reference/ois/next/processing.md:122]: Changed to `The maximum duration` in all places.

File: docs/reference/ois/next/processing.md (hunk 13 above)
Comment (Siegrift) [reply docs/reference/ois/next/processing.md]: Thanks. I also double checked that the link works on the preview site (https://vitepress-docs--pr669-processing-v2-d3oh59pa.web.app/reference/airnode/latest/understand/http-gateways.html#http-signed-data-gateway).
